plutus-tx/src/Language/PlutusTx/IsData.hs

```diff
@@ -78,12 +79,9 @@ instance (IsData a, IsData b) => IsData (Either a b) where
     fromData (Constr i [bd]) | i == 1 = Left <$> fromData bd
     fromData _               = Nothing
 
--- TODO: maybe this should actually be encoded as a list!
 instance IsData a => IsData [a] where
     {-# INLINABLE toData #-}
-    toData [] = Constr 0 []
-    toData (x:xs) = Constr 1 [toData x, toData xs]
+    toData xs = Seq (fmap toData xs)
```

Comment:
It's kind of nice to have a separate constructor of `Data`, but I'd probably just define it as

```haskell
toData xs = Constr 0 $ map toData xs
```

(and `fromData` accordingly). That `Constr` is not self-describing when it comes to other data types (i.e. its actual meaning depends on the data type it's being converted to/from) and so I wouldn't make lists any more special in this respect.

Reply:
Although I admit it's kind of a hack.

Reply:
Hm, the semantics are slightly off then. `Constr` is really supposed to have a tuple of arguments of some known size, so using it for variable length lists is a bit odd.

Note that it gets turned into almost the same thing in CBOR: a `Constr` is just a tagged list, a `Seq` is an untagged list. So the difference is mainly how it gets represented on the Plutus side.

Reply:
> Hm, the semantics are slightly off then. Constr is really supposed to have a tuple of arguments of some known size, so using it for variable length lists is a bit odd.

That's why I say it's a hack.

> Note that it gets turned into almost the same thing in CBOR: a Constr is just a tagged list, a Seq is an untagged list. 

Which is the reason I propose we can just use an empty tag.

Either way is fine, I just wanted to mention this option.

Reply:
Yeah, it's not a bad idea I think I just like this one slightly more.